### PR TITLE
[NEW] Youtube Broadcasting

### LINKED ISSUE #9543: Livestream & Broadcasting on Rocket.Chat
## Description:

This feature integrates video livestream and broadcasting to Rocket.Chat servers. When enabled, the Livestream tab open the possibility of adding embedded video links to a channel which every user can access and watch through the new popout component. The main objective of the feature is enable a better communication within streaming communities, we also are working with Brave and BAT to enhance the compatibility with the platform so content creators are rewarded not only by creating a better community but for their efforts as well through Basic Attention Token.

The first phase of the integration is focused on the channel livestreaming platform, creation of the popout component that enables watching a livestream while changing conversation channel within Rocket.Chat.

The second phase of the integration is related to the broadcasting interface and how users can broadcast directly from their Rocket.Chat instance once connected with an OAuth provider (Youtube will be used at first), this will also improve the Livestream functionalities by having automated updates from the broadcaster to selected channels.

### Phase 1 - Livestream UI ([Pull Request](https://github.com/RocketChat/Rocket.Chat/pull/9255))

- [x] Admin settings to enable Livestream feature
- [x] Livestream tab on channel contextual sidebar
- [x] Enable admins to set URL and streaming mode (audio or video)
- [x] Parse different URLs to create a embeddable link for the livestream (Youtube for now)
- [x] Create Popout component following the design provided
- [x] Create modal window that persists video through a change of channels
- [x] Enable users to watch livestream through the popout component
- [x] Allow users to move the popout by drag-n-drop
- [x] Allow users to minimize the video without it being stopped
- [x] Allow users to play/pause/mute the livestream (only for himself)

### Phase 2 - Broadcasting UI

- [ ] Admin settings to enable and configure broadcasting settings
- [ ] Extend Livestream channel sidebar
- [ ] Integrate youtube streaming API into Rocket.Chat 
- [ ] Enable users with permission to broadcast to their youtube channel (once configured in the admin panel)

### LINKED ISSUE #10070: liveStreamTab.css:40:5: variable '--rc-color-primary-lightest' and '--rc-color-primary-dark'  is undefined and used without a fall
### Description:
There was an error when I built the RocketChat environment locally. What should I do?

postcss-custom-properties: /home/luo/github/Rocket.Chat/packages/rocketchat-livestream/client/styles/liveStreamTab.css:19:5: variable '--rc-color-primary-lightest' is undefined and used without a fallbackpostcss-custom-properties: /home/luo/github/Rocket.Chat/packages/rocketchat-livestream/client/styles/liveStreamTab.css:40:5: variable '--rc-color-primary-dark' is undefined and used without a fall

### Server Setup Information:

* Version of Rocket.Chat Server: 0.63
* Operating System: 
* Deployment Method(snap/docker/tar/etc): localhost
* Number of Running Instances: 
* DB Replicaset Oplog: 
* Node Version: 
* mongoDB Version:

### PR DESCRIPTION
**Pull request by gdelavald**
@RocketChat/core 

Closes #9543
Closes #10070 

Link to documentation PR: https://github.com/RocketChat/docs/pull/746

### Youtube Broadcasting
This new feature adds an integration between Rocket.Chat and Youtube Live, when enabled and set up (check configuration details below) your channel owners will be able to broadcast their camera feed live to the Livestream feature (check #9543 for details) which allows all users on the channel to watch a common stream.

#### Configuration
The Broadcasting feature can only be used with the Livestream Tab feature, check the admin panel for Livestream & Broadcasting for more details, currently the admin needs a valid Google app with access to the Youtube API (More details on the Google Developer page), this will allow the administrator to have a valid API Key, Client ID & Secret which are used on the Broadcasting feature.

#### Usage
After a correct set up, the Livestream tab will now allow you to Broadcast your camera when editing the source url.
When opening the popout, it might take a moment while we prepare the broadcast and livestream for your session, this will be indicated by the blinking icon and "Preparing" status displayed in the control bar. After that initial phase, the broadcast will be `Ready` and available to start with your click.

rxsmgjkz6v
Once the broadcasting is started, the users will be able to watch you live through the livestream tab.

The broadcasting can be finished with another click on the red colored circle icon or by closing the popout window. Once it is finished, users will be able to check the video recording also.

Thanks @ggazzo for all the hard work on this.

### CODE REVIEW

**Comment by gdelavald** [comment]
@ggazzo updated the problems you pointed, please review again

File: packages/rocketchat-livestream/package.js [hunk 1]
```diff
@@ -13,16 +13,29 @@ Package.onUse(function(api) {
...
 		'server/functions/saveStreamingOptions.js',
 		'server/settings.js'
 	], 'server');
+
+	Npm.depends({
+		'googleapis': '25.0.0',
+		'fluent-ffmpeg': '2.1.2',
+		'websocket-stream': '5.1.2',
+		'ws': '5.0.0',
+		'express': '4.15.4'
```
Comment (karlprieb) [inline packages/rocketchat-livestream/package.js]: review dependencies <3

File: packages/rocketchat-livestream/server/routes.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	get: function functionName() {
+		const { code, state } = this.queryParams;
+
+		const { userId } = JSON.parse(state);
+
+		const clientAuth = new OAuth2(RocketChat.settings.get('Broadcasting_client_id'), RocketChat.settings.get('Broadcasting_client_secret'), `${ RocketChat.settings.get('Site_Url') }/api/v1/livestream/oauth/callback`);
+
+		const ret = Meteor.wrapAsync(clientAuth.getToken.bind(clientAuth))(code);
+
+		console.log(ret);
```
Comment (karlprieb) [inline packages/rocketchat-livestream/server/routes.js]: this `console.log` is important to have?

Comment (gdelavald) [reply packages/rocketchat-livestream/server/routes.js]: Nope, removing.

**Comment by karlprieb** [comment]
@ggazzo yep! I will wait @gdelavald fix the order tabBar order to approve this PR.

**Comment by IQ2022** [comment]
cool idea, but what about broadcasting your chat? or maybe desktop/ slide deck with Camera similar to doing a tutorial via Camtasia

**Comment by gdelavald** [comment]
@IQ2022 improvements and enhancements are always a possibility. This fit our first plan to broadcast your camera, new sources can be added later with the help of the community.

**Comment by Syirrus** [comment]
Are their step by step instruction on how to set this up?

**Pull request merged**